File: LeetCode/1420. Build Array Where You Can Find The Maximum Exactly K Comparisons/solution.py
from functools import lru_cache
class Solution(object):
    def numOfArrays(self, N, M, K):
        MOD = 10**9 + 7
        
        @lru_cache(None)
        def dp(i=0, m=-1, rem=K):
            if i == N:
                return +(rem == 0)
            if rem < 0:
                return 0

            ans = 0
            for a in range(1, M + 1):
                if a > m: #newmax
                    ans += dp(i+1, a, rem - 1)
                    ans %= MOD
                else:
                    ans += dp(i+1, m, rem)
                    ans %= MOD
            return ans

        return dp()

DP.  dp(i, m, rem) means that you have to write the array arr[i:] where we had a max of m = max(a[:i]) and 
require a search cost of rem.

# from functools import lru_cache
# class Solution:
#     def numOfArrays(self, N: int, M: int, K: int) -> int:
#         @lru_cache(None)
#         def dfs(m,n,k):
#             if n == 0:
#                 if k == 0:
#                     return 1
#                 else:
#                     return 0
#             if m < 1 or k < 0:
#                 return 0
#             sums = 0
#             for x in range(1,m+1):
#                 sums += dfs(x-1,n-1,k-1)
#                 sums += dfs(m,n-1,k)-dfs(x-1,n-1,k)
#             return sums%(10**9+7)
#         return dfs(M,N,K)

File: LeetCode/1420. Build Array Where You Can Find The Maximum Exactly K Comparisons/solution.cpp
using ll = long long;

const int MOD = 1e9+7;

class Solution {
public:
    int numOfArrays(int n, int m, int k) {
        vector<vector<int>> dp(m+1,vector<int>(k+1,0));
        dp[0][0] = 1;
        for (int len = 1; len <= n; len++){
            vector<vector<int>> ndp(m+1,vector<int>(k+1,0));
            for (int kk = 1; kk <= min(len,k); kk++){
                int lo = 0;
                
                for (int mm = 1; mm <= m; mm++){
                    lo = (lo+dp[mm-1][kk-1])%MOD;
                    ndp[mm][kk] = ((ll)dp[mm][kk]*mm+lo)%MOD;
                    // cout << len << " " << kk << " " << mm << " " << ndp[mm][kk] << endl;
                }
            }
            swap(dp,ndp);
        }
        int sums = 0;
        for (int mm = 1; mm <= m; mm++)
            sums = (sums+dp[mm][k])%MOD;
        return sums;
    }
};






// using ll = long long;

// const int MOD = 1e9+7;
// const int maxm = 105, maxk = 105;

// int dp[maxm][maxk], ndp[maxm][maxk];

// class Solution {
// public:
//     int numOfArrays(int n, int m, int k) {
//         memset(dp, 0, sizeof(int)*maxm*maxk);
//         dp[0][0] = 1;
//         for (int len = 1; len <= n; len++){
//             memset(ndp, 0, sizeof(int)*maxm*maxk);
//             for (int kk = 1; kk <= min(len,k); kk++){
//                 int lo = 0;
                
//                 for (int mm = 1; mm <= m; mm++){
//                     lo = (lo+dp[mm-1][kk-1])%MOD;
//                     ndp[mm][kk] = ((ll)dp[mm][kk]*mm+lo)%MOD;
//                     // cout << len << " " << kk << " " << mm << " " << ndp[mm][kk] << endl;
//                 }
//             }
//             swap(dp,ndp);
//         }
//         int sums = 0;
//         for (int mm = 1; mm <= m; mm++)
//             sums = (sums+dp[mm][k])%MOD;
//         return sums;
//     }
// };


using ll = long long;

const int MOD = 1e9+7;
const int maxm = 105, maxk = 105;

int dp[maxk][maxm], ndp[maxk][maxm];

class Solution {
public:
    int numOfArrays(int n, int m, int k) {
        memset(dp, 0, sizeof(int)*maxm*maxk);
        dp[0][0] = 1;
        for (int len = 1; len <= n; len++){
            memset(ndp, 0, sizeof(int)*maxm*maxk);
            for (int kk = 1; kk <= min(len,k); kk++){
                int lo = 0;
                
                for (int mm = 1; mm <= m; mm++){
                    lo = (lo+dp[kk-1][mm-1])%MOD;
                    ndp[kk][mm] = ((ll)dp[kk][mm]*mm+lo)%MOD;
                    // cout << len << " " << kk << " " << mm << " " << ndp[mm][kk] << endl;
                }
            }
            swap(dp,ndp);
        }
        int sums = 0;
        for (int mm = 1; mm <= m; mm++)
            sums = (sums+dp[k][mm])%MOD;
        return sums;
    }
};